### PR TITLE
DataViews: Update template parts view

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Part of: https://github.com/WordPress/gutenberg/issues/55083

This PR updates the template parts list in site editor to use DataViews. Since templates and template parts have almost identical handling I've created a thin DataViews wrapper for both and most changes are checks about the type to use some different labels. 

## Testing Instructions
1. Test template parts page
2. Templates page should work as before and needs to be tested too, because of the thin wrapper

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/16275880/c1500d0e-6c47-4315-9e93-2695c37fa388

### CODE REVIEW

File: packages/edit-site/src/components/page-templates/index.js [hunk 1]
```diff
@@ -1,424 +1,10 @@
...
-							</div>
-						) }
-					</div>
-				</Page>
-			) }
-		</>
-	);
+export default function PageTemplates() {
+	return <DataviewsTemplatesTemplateParts postType={ TEMPLATE_POST_TYPE } />;
 }
```
Comment (youknowriad) [inline packages/edit-site/src/components/page-templates/index.js]: Is this file still necessary, I think if you remove it you may end up with a better "git diff". Also maybe you should rename the folder since it now contains the reusable component.

File: packages/edit-site/src/components/page-template-parts/dataviews-template-parts.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Internal dependencies
+ */
+import { TEMPLATE_PART_POST_TYPE } from '../../utils/constants';
+
+import DataviewsTemplatesTemplateParts from '../page-templates/dataviews-templates-template-parts';
+
+export default function DataviewsTemplateParts() {
+	return (
+		<DataviewsTemplatesTemplateParts postType={ TEMPLATE_PART_POST_TYPE } />
```
Comment (youknowriad) [inline packages/edit-site/src/components/page-template-parts/dataviews-template-parts.js]: Is this file really needed, why not just use the underlying component directly?

Comment (ntsekouras) [reply packages/edit-site/src/components/page-template-parts/dataviews-template-parts.js]: For now it's not, but I thought it could when we add the view handling.. I can definitely remove for now.

**Comment by oandregal** [comment]
There's an issue with the details: it redirects to the wrong postType. 

https://github.com/WordPress/gutenberg/assets/583546/842bf489-8929-405b-94f5-ee565b48eb70

Additionally, in implementing Template Parts, the current approach to naming shows as something to improve. A simple tweak could be to update the copy of the button to say "Add new".

Captura de ecrã 2024-01-19, às 10 29 05

Unrelated to this PR, but I've noticed the new dataviews-powered Patterns screen doesn't use the "Add new" button in the content area, it still uses the `+` icon in the sidebar. 

Captura de ecrã 2024-01-19, às 10 32 34

File: packages/edit-site/src/components/page-templates-template-parts/index.js [hunk 3]
```diff
@@ -248,35 +266,36 @@ export default function DataviewsTemplates() {
 				maxWidth: 400,
 				minWidth: 320,
 				enableSorting: false,
+			} );
+		}
+		// TODO: This should be improved when we implement field ordering with an API.
```
Comment (oandregal) [inline packages/edit-site/src/components/page-templates-template-parts/index.js]: What does this mean? Can you expand on this?

File: packages/edit-site/src/components/page-templates-template-parts/index.js [hunk 4]
```diff
@@ -92,19 +98,24 @@ function normalizeSearchInput( input = '' ) {
...
-
+	const linkProps = {
+		params: {
+			postId: item.id,
+			postType: item.type,
+			canvas: 'edit',
+		},
+	};
+	if ( item.type === TEMPLATE_PART_POST_TYPE ) {
+		linkProps.state = {
```
Comment (oandregal) [inline packages/edit-site/src/components/page-templates-template-parts/index.js:107]: I haven't seen this in use, what is it for? I presume it's for setting the path in case the users click the browser's back button after they've gone to the editor by clicking the title. Is this correct? If so, why templates don't need this?

File: packages/edit-site/src/components/page-templates-template-parts/actions.js [hunk 5]
```diff
@@ -167,8 +167,15 @@ export const deleteTemplateAction = {
 export const renameTemplateAction = {
 	id: 'rename-template',
 	label: __( 'Rename' ),
-	isEligible: ( template ) =>
-		isTemplateRemovable( template ) && template.is_custom,
+	isEligible: ( template ) => {
+		if (
+			! isTemplateRemovable( template ) ||
```
Comment (oandregal) [inline packages/edit-site/src/components/page-templates-template-parts/actions.js:174]: I found this logic difficult to grok. What would you think of rewriting it by stating the conditions for each post type independently? Something along these lines could work:

```js
if (
    template.type === TEMPLATE_PART_POST_TYPE &&
    isTemplateRemovable( template )
) {
    return true;
}
if (
    template.type === TEMPLATE_POST_TYPE &&
    isTemplateRemovable( template) &&
    template.is_custom
) {
    return true;
}

return false;
```

File: packages/edit-site/src/store/private-actions.js [hunk 6]
```diff
@@ -80,20 +86,22 @@ export const removeTemplates =
...
 				successMessage = sprintf(
 					/* translators: The template/part's name. */
 					__( '"%s" deleted.' ),
-					decodeEntities( templateTitle )
+					decodeEntities( title )
 				);
 			} else {
-				successMessage = __( 'Templates deleted.' );
+				successMessage = isTemplate
+					? __( 'Templates deleted.' )
```
Comment (oandregal) [inline packages/edit-site/src/store/private-actions.js:103]: I've mentioned this elsewhere, but I think we could do with using a generic language. For example, `Items deleted.`.

The users already have the context of what they are – they know whether they are working with templates or template parts, don't need reminding. And this code would be a lot simpler.

I don't have a strong opposition to doing it this way, and it works fine. I just want to share an alternative that I find clearer for everyone.

File: packages/edit-site/src/components/page-templates-template-parts/index.js (hunk 3 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-templates-template-parts/index.js]: The plan is to support fields reordering, which would require an API like `order` or something. Right now I'm creating  the array -> adding a conditional field in the wanted order (3rd) -> add another one after that. With ordering API I believe we wouldn't need to construct the fields array like that.

File: packages/edit-site/src/components/page-templates-template-parts/index.js [hunk 7]
```diff
@@ -1,3 +1,7 @@
+/**
+ * This component is a thin DataViews wrapper around templates and template parts.
```
Comment (oandregal) [inline packages/edit-site/src/components/page-templates-template-parts/index.js]: This is me nitpicking. When I read "thin wrapper" in the PR description I was intrigued. Now that I see it's a React component (that takes props), I find it a bit too fancy 😇

**Review by oandregal — COMMENTED** [review]
I found an issue to fix ([details](https://github.com/WordPress/gutenberg/pull/57952#issuecomment-1900062315)) and left a couple of comments for clarification. Other than that, this is working great 👏

File: packages/edit-site/src/components/page-templates-template-parts/index.js (hunk 3 above)
Comment (oandregal) [reply packages/edit-site/src/components/page-templates-template-parts/index.js]: This is better as a comment :)

File: packages/edit-site/src/store/private-actions.js (hunk 6 above)
Comment (ntsekouras) [reply packages/edit-site/src/store/private-actions.js:103]: I agree  with this and would need less code :). Le't handle this in a follow up though because others might chip in about the clarity of the messages and wouldn't want to block  this PR for that.

**Review by oandregal — APPROVED** [review]
